### ISSUE TITLE
expandable_segments with PYTORCH_CUDA_ALLOC_CONF reduces VRAM

### ISSUE THREAD

**Issue by winglian** [pytorch/torchtune#1185]
In [Unsloth](https://github.com/unslothai/unsloth/blob/main/unsloth/__init__.py#L53-L54):
```python
# Reduce VRAM usage by reducing fragmentation
os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"
```

using the defaults in torchtune:

`tune run lora_finetune_single_device --config llama3/8B_qlora_single_device`

- A100 80GB
- Measure @ 20steps 
- PYTORCH_CUDA_ALLOC_CONF="expandable_segments:True"

We should dig into why this provides such a large memory improvement and consider making it the default setting, or documenting it somewhere if not.

baseline: 16.39GiB @ 21.6sec/it
w/ compile: 16.63GiB @ 8.5sec/it
w/ expandable_segments: 10.831GiB @ 20.56sec/it
w/ compile + expandable_segments:  10.919GiB @ 8.5sec/it

**Comment by felipemello1** [comment]
I wonder if we can squeeze more from it, and make it model + machine specific. E.g. in the first pass of the training, run a scrip that optimizes for it, since there are a few parameters associated with this flag

```
    os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "backend:native,\
                                            max_split_size_mb:1024,\
                                            roundup_power2_divisions:8,\
                                            roundup_bypass_threshold_mb:256,\
                                            garbage_collection_threshold:0.8"
```

https://www.educative.io/answers/memory-management-using-pytorchcudaallocconf
https://dev.to/shittu_olumide_/how-can-i-set-maxsplitsizemb-to-avoid-fragmentation-in-pytorch-37h9

**Comment by pbontrager** [comment]
Adding a comment to track this discussion in Pytorch core https://github.com/pytorch/pytorch/issues/130330

**Comment by felipemello1** [comment]
I did some tests. It doesnt impact allocated memory, only reserved memory (cache). Pre-allocating memory before training starts has the same result. To pre-allocate, we create a batch with max_seq_len, fun a forward/backward, but no step. 

If we dont do it, as batch_sizes change, the memory gets fragmented, resulting in higher reserved memory. But it didnt seem harmful. I will put a PR suggesting we do it in our recipes. Worst case scenario, the user will get OOM in the very first batch, instead in the middle of training.

![image](https://github.com/user-attachments/assets/ecee0930-b1e3-4b61-86ed-2c4afd4fd945)

**Comment by ebsmothers** [comment]
@winglian where did we land on this? The numbers here are for reserved memory, right? Do we see similar improvements on peak allocated memory as well?

**Comment by felipemello1** [comment]
In my experiments, it impacts only reserved memory, and only when we dont pre-allocate memory for the batch (create a dummy batch with max_seq_len, run forward/backward, delete). Since batch_n+1 can be larger than batch_n, then memory needs to be reallocated, causing higher reserved memory.

In the example below, I am using packed dataset, so every batch has max_seq_len. No significant difference with/without flag.

<img width="1214" alt="image" src="https://github.com/user-attachments/assets/53ab4b42-8cee-4523-99ce-ea6928328929">

**Comment by felipemello1** [comment]
But apparently this can cause OOMs? 

<img width="1576" alt="image" src="https://github.com/user-attachments/assets/02b90139-5d36-42f9-81ee-3861ada4948a">

**Comment by pbontrager** [comment]
I think the prudent thing to would be to wait for this to land in core with all the necessary safety updates that go with it, and not enable this flag ourself.

**Comment by felipemello1** [comment]
Just wanted to confirm that running on A100, with the flag i can run bs=4, but without, its OOMs

![image](https://github.com/user-attachments/assets/d92461ba-63af-4f5e-a385-a617a0d66a71)

```
tune run --nproc_per_node 8 lora_finetune_distributed --config llama3/8B_lora \
batch_size=4 \
model.lora_attn_modules="['q_proj', 'v_proj']" \
model.apply_lora_to_mlp=False \
model.apply_lora_to_output=False \
metric_logger.name=LoRA__8192__nproc8__bsz4__ActCkptTrue \
dataset.max_seq_len=8192 \
dataset.source=Yukang/LongAlpaca-12k \
dataset.packed=False \
dataset.split="train[:10%]" \
metric_logger=torchtune.utils.metric_logging.WandBLogger \
metric_logger.project=mem_prealloc \
gradient_accumulation_steps=1 \
log_every_n_steps=1 \
log_peak_memory_stats=True \
max_steps_per_epoch=20 \
epochs=1 \
compile=False \
enable_activation_checkpointing=True
```

**Comment by joecummings** [comment]
> Just wanted to confirm that running on A100, with the flag i can run bs=4, but without, it OOMs

This would imply that we should be paying attention to reserved attention too, right? @ebsmothers

**Comment by ebsmothers** [comment]
> > Just wanted to confirm that running on A100, with the flag i can run bs=4, but without, it OOMs
> 
> This would imply that we should be paying attention to reserved attention too, right? @ebsmothers

Sorry I missed this before. But based on @felipemello1's screenshot I would say that's not the case. It seems to me like this flag also impacts both active and allocated memory. In that case it seems pretty clear to me that we should enable it

**Comment by felipemello1** [comment]
> It looks like this flag also impacts both active and allocated memory

No, sorry, thats misleading. Green has bsz=3. When i go to bsz=4, it doesnt work, unless it use the flag.

It does NOT impact allocated memory, only reserved

**Comment by ebsmothers** [comment]
Thanks @felipemello1 for clarifying, I should've looked at the legend more closely. 

So for actual next steps here: it seems like there are cases where anecdotally this does prevent OOMs. However there is some feature gap around CUDA IPC (as discussed in the issue linked by @pbontrager) preventing core from enabling it by default at this exact moment. We can wait for that, but if we want to unblock this sooner, I would say we can:

(1) enable this config on a branch, run on a broader set of recipes to ensure no obvious breakages or huge perf regressions.

Depending on the result of that, either:

(2a) enable this by default for all our recipes, or
(2b) gate it behind a config.

In my mind (2b) is not the ideal outcome cause it's just one extra bespoke field to worry about. But open to discussing once we have more comprehensive results here.

**Comment by Seedmanc** [comment]
What's the support matrix for expandable segments? I keep getting denied with "not supported on this platform" using EasyDiffusion and EasyTraining GUIs, but it's no stating whether I'm missing hardware or software support.
I'm running Cuda 12.1 on Win10x64 and RTX 2070S 8Gb, torch-2.3.1+cu121, torchvision-0.18.1+cu121.

**Comment by felipemello1** [comment]
@Seedmanc , seems related: https://github.com/pytorch/pytorch/issues/122057

**Comment by Petros626** [comment]
@felipemello1 random questiobn: how did you custom named the `wandb` logs?

**Comment by felipemello1** [comment]
> @felipemello1 random questiobn: how did you custom named the `wandb` logs?

@Petros626 
```
tune run --nproc_per_node 2 full_finetune_distributed --config llama3_2/3B_full \
metric_logger=torchtune.training.metric_logging.WandBLogger \
metric_logger.project=profiling \
metric_logger.tags=[my_experiment_name] \
metric_logger.name=my_log_name 
```

**Comment by ebsmothers** [comment]
I am gonna close this issue as it's quite old and we ultimately decided to have PyTorch handle this

**Comment by kissint8** [comment]
> Just wanted to confirm that running on A100, with the flag i can run bs=4, but without, it OOMs
> 
> editing for clarity: Green has bsz=3. When i go to bsz=4, it doesnt work, unless I use the flag. The info is in the run name. It does NOT impact allocated memory, only reserved.
> 
> ![image](https://github.com/user-attachments/assets/d92461ba-63af-4f5e-a385-a617a0d66a71)
> 
> ```
> tune run --nproc_per_node 8 lora_finetune_distributed --config llama3/8B_lora \
> batch_size=4 \
> model.lora_attn_modules="['q_proj', 'v_proj']" \
> model.apply_lora_to_mlp=False \
> model.apply_lora_to_output=False \
> metric_logger.name=LoRA__8192__nproc8__bsz4__ActCkptTrue \
> dataset.max_seq_len=8192 \
> dataset.source=Yukang/LongAlpaca-12k \
> dataset.packed=False \
> dataset.split="train[:10%]" \
> metric_logger=torchtune.utils.metric_logging.WandBLogger \
> metric_logger.project=mem_prealloc \
> gradient_accumulation_steps=1 \
> log_every_n_steps=1 \
> log_peak_memory_stats=True \
> max_steps_per_epoch=20 \
> epochs=1 \
> compile=False \
> enable_activation_checkpointing=True
> ```


I am very interested in your memory image. Is there any tool can get this visualization result, OR,  which function to get these data?

**Comment by felipemello1** [comment]
@kissint8 , this is weights and biases. In the command, it is defined by the logger "metric_logger=torchtune.utils.metric_logging.WandBLogger"
